gdpr/vendorlist-fetching.go

```diff
@@ -75,13 +78,38 @@ func preloadCache(ctx context.Context, client *http.Client, urlMaker func(uint16
 			firstListVersion: 1,
 		},
 	}
-	for _, v := range versions {
-		latestVersion := saveOne(ctx, client, urlMaker(v.specVersion, 0), saver)
+	var wgLatestVersion errgroup.Group
+	var wgSpecificVersion errgroup.Group
 
-		for i := v.firstListVersion; i < latestVersion; i++ {
-			saveOne(ctx, client, urlMaker(v.specVersion, i), saver)
-		}
+	if conf.MaxConcurrencyInitFetchLatestVersion > 0 {
+		wgLatestVersion.SetLimit(conf.MaxConcurrencyInitFetchLatestVersion)
+	}
+
+	if conf.MaxConcurrencyInitFetchSpecificVersion > 0 {
+		wgSpecificVersion.SetLimit(conf.MaxConcurrencyInitFetchSpecificVersion)
 	}
+
+	tsStart := time.Now() // For logging how long this takes
+	for _, v := range versions {
+		specVersion := v.specVersion
+		firstVersion := v.firstListVersion
+		wgLatestVersion.Go(func() error {
+			latestVersion := saveOne(ctx, client, urlMaker(specVersion, 0), saver)
+			for i := firstVersion; i < latestVersion; i++ {
+				currentVersion := i
+				wgSpecificVersion.Go(func() error {
+					saveOne(ctx, client, urlMaker(specVersion, currentVersion), saver)
+					return nil
+				})
+			}
+			return nil
+		})
+	}
+
+	wgLatestVersion.Wait()
+	wgSpecificVersion.Wait()
```

Comment:
Hi !

I don't know if it matters. 

The WaitGroup `wgSpecificVersion` will likely wait for more Goroutines than `wgLatestVersion`.

Globally, you start Goroutines with a limit of `conf.MaxConcurrencyInitFetchLatestVersion + conf.MaxConcurrencyInitFetchSpecificVersion` concurrent Go routines.

So why have 2 wait groups instead of just 1 wait group ?

Reply:
@sebhtml , good thought, however the reason i did it is to have fine granularity control of two different behaviours:

1. fetching the latest version
2. fetching each specific version.
3. provide sequential execution as default behaviour so the decision of tuning the controlling parameters are determined by the host.


by using only one wait group, there are two thoughts behind it

1. `conf.MaxConcurrencyInitFetchLatestVersion` and `conf.MaxConcurrencyInitFetchSpecificVersion` loses its meaning since one wait group wont be able to control two types of behaviour precisely.

2. lets say you use one wait group and the corresponding config is `conf.MaxConcurrencyInitFetch` , and then you set the value to `1` ( which is supposed to mean it is a sequential execution )  with the following implementation 
```
var wg errorgroup.Group
for _, v := range versions {
		specVersion := v.specVersion
		firstVersion := v.firstListVersion
		wg.Go(func() error {
			latestVersion := saveOne(ctx, client, urlMaker(specVersion, 0), saver)
			for i := firstVersion; i < latestVersion; i++ {
				currentVersion := i
				wg.Go(func() error {
					saveOne(ctx, client, urlMaker(specVersion, currentVersion), saver)
					return nil
				})
			}
			return nil
		})
	}
```
then you have deadlock

Reply:
Oh I see. You have a good point. I therefore resolve this thread.